File: Leetcode_Python/validate_binary_search_tree.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return a boolean
    def isValidBST(self, root):
        if root == None: return True
        lis = self.inOrder(root, [])
        if len(lis) <= 1:  return True
        for i in xrange(len(lis) - 1):
            if lis[i] >= lis[i + 1]:
                return False
        return True

    def inOrder(self, node, re):
        if node == None:  return
        if node.left:
            self.inOrder(node.left, re)
        re.append(node.val)
        if node.right:
            self.inOrder(node.right, re)
        return re

File: cleanCode50/ValidateBinarySearchTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public boolean isValidBST(TreeNode root) {
        return validBST(root, Double.NEGATIVE_INFINITY, Double.POSITIVE_INFINITY);
    }
    
    private boolean validBST(TreeNode root, Double min, Double max) {
        if (root == null) {
            return true;
        } else if (root.val > min && root.val < max) {
            Double rootValue = Double.valueOf(root.val);
            return validBST(root.left, min, rootValue) && validBST(root.right, rootValue, max);
        } else {
            return false;
        }
    }
}